### PR TITLE
feat: add scalar decomposition and fast EC simultaneous multiplication

### LINKED ISSUE #7: Implement generic scalar decomposition as generic method
@mariocao commented on [Mon Aug 12 2019](https://github.com/witnet/research/issues/61)

Integrate scalar decomposition support into the `EllipticCurve` contract:

- [x] Implement scalar decomposition
- [x] Tests

### PR DESCRIPTION
**Pull request by mariocao**
This PR includes a new contract that implements:

- Scalar decomposition
- wNAF scalar representation
- fast EC simultaneous multiplication using wNAF

The simultaneous multiplication computes 2 EC multiplications of the form `kP + lQ`, where:

- `k` and `l` are scalars
- `P` and `Q` are EC points in affine coordinates

Closes #7

### CODE REVIEW

File: contracts/EllipticCurve.sol [hunk 1]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+  ) internal pure {
+    uint256 p = fieldOrder;
+    uint256 beta = 0x7ae96a2b657c07106e64479eac3434e99cf0497512f58995c1396c28719501ee;
+    
+    uint256[3][4] memory iPj;
+    uint256[3] memory double;
+
+    // P1 Lookup Table
+    iPj = iP[0];
+    iPj[0] = [P_Q[0], P_Q[1], 1];  						// P1
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: The comment is too far on the right side.

File: contracts/EllipticCurve.sol [hunk 2]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+    iP[1][0] = [mulmod(beta, P_Q[0], p), P_Q[1], 1];	// P2
+
+    iP[1][1] = [mulmod(beta, iPj[1][0], p), iPj[1][1], iPj[1][2]];
+    iP[1][2] = [mulmod(beta, iPj[2][0], p), iPj[2][1], iPj[2][2]];
+    iP[1][3] = [mulmod(beta, iPj[3][0], p), iPj[3][1], iPj[3][2]];
+
+    // Q1 Lookup Table
+    iPj = iP[2];
+    iPj[0] = [P_Q[2], P_Q[3], 1];   
+                    	// Q1
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: Misplaced comment

File: contracts/EllipticCurve.sol [hunk 3]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+        if and(k, 1) {  // if k is odd:
+          ki := mod(k, 16)
+          k := add(sub(k, ki), mul(gt(ki, 8), 16))
+          // if sign = 1, store ki; if sign = -1, store 16 - ki
+          mstore8(add(ptr, length), add(mul(ki, sign), sub(8, mul(sign, 8))))
+        }
+        length := add(length, 1)
+        k := div(k, 2)
+      }
+    //log3(ptr, 1, 0xfabadaacabada, d, length)
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: What's the point of this line?

File: contracts/EllipticCurve.sol [hunk 4]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+  /// @notice Computes the WNAF representation of an integer, and puts the resulting array of coefficients in memory
+  /// @param d A 256-bit integer
+  /// @return (ptr, length) The pointer to the first coefficient, and the total length of the array
+  function _wnaf(int256 d) internal pure  returns (uint256 ptr, uint256 length) {
+    
+    int sign = d < 0 ? -1 : int(1);
+    uint256 k = uint256(sign * d);
+
+    length = 0;
+    assembly
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: Do we have a benchmark of this block of code in Solidity vs. Assembly?

File: contracts/EllipticCurve.sol [hunk 5]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+    uint256[3] memory Q
+  ) internal pure returns(bool) {
+    uint256 p = fieldOrder;
+
+    uint256 Qz = Q[2];
+    uint256 Pz = P[2];
+    if(Pz == 0){
+      return Qz == 0;   // P and Q are both zero.
+    } else if(Qz == 0){
+      return false;       // Q is zero but P isn't.
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: Please align comments.

File: contracts/EllipticCurve.sol [hunk 6]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+    }
+
+    // Now we're sure none of them is zero
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]:
```suggestion
     // At this point, we are certain that `Pz` and `Qz` are not zero
```

File: contracts/EllipticCurve.sol [hunk 7]
```diff
@@ -251,6 +252,316 @@ contract EllipticCurve {
...
+  /// @param k_l An array with the decomposition of k and l values, i.e., [k1, k2, l1, l2]
+  /// @param P_Q An array with the affine coordinates of both P and Q, i.e., [P1, P2, Q1, Q2]
+  function _sim_mul(
+    int256[4] memory k_l, 
+    uint256[4] memory P_Q,
+    uint256 a,
+    uint256 pp
+  ) public pure returns (uint[3] memory Q) {
+
+    /*require(
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: What's the point of this code block being commented out?

File: contracts/EllipticCurve.sol [hunk 8]
```diff
@@ -268,7 +579,7 @@ contract EllipticCurve {
     uint256 y2,
     uint256 z2,
     uint256 pp)
-  internal pure returns (uint256 qx, uint256 qy, uint256 qz)
+  internal pure  returns (uint256 qx, uint256 qy, uint256 qz)
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: Redundant whitespace.

File: contracts/EllipticCurve.sol [hunk 9]
```diff
@@ -367,7 +678,7 @@ contract EllipticCurve {
     uint256 z,
     uint256 a,
     uint256 pp)
-  internal pure returns (uint256 qx, uint256 qy, uint256 qz)
+  internal pure  returns (uint256 qx, uint256 qy, uint256 qz)
```
Comment (aesedepece) [inline contracts/EllipticCurve.sol]: Redundant whitespace.

File: contracts/EllipticCurve.sol (hunk 1 above)
Comment (mariocao) [reply contracts/EllipticCurve.sol]: This part of the code was removed :)

**Pull request merged**